src/command.cpp

```diff
@@ -603,62 +270,25 @@ bool DoCommandP(TileIndex tile, uint32 p1, uint32 p2, uint32 cmd, CommandCallbac
 }
 
+/** Helper to make a desync log for a command. */
+void DoCommandHelperBase::LogCommandExecution(Commands cmd, StringID err_message, TileIndex tile, const CommandDataBuffer &args, bool failed)
+{
+	Debug(desync, 1, "{}: {:08x}; {:02x}; {:02x}; {:08x}; {:08x}; {:06x}; {} ({})", failed ? "cmdf" : "cmd", _date, _date_fract, (int)_current_company, cmd, err_message, tile, FormatArrayAsHex(args), GetCommandName(cmd));
+}
 
 /**
- * Helper to deduplicate the code for returning.
- * @param cmd   the command cost to return.
+ * Prepare for the test run of a command proc call.
+ * @param cmd_flags Command flags.
+ * @param tile Tile of command execution.
+ * @param[in,out] cur_company Backup of current company at start of command execution.
+ * @return True if test run can go ahead, false on error.
  */
-#define return_dcpi(cmd) { _docommand_recursive = 0; return cmd; }
-
-/*!
- * Helper function for the toplevel network safe docommand function for the current company.
- *
- * @param tile The tile to perform a command on (see #CommandProc)
- * @param p1 Additional data for the command (see #CommandProc)
- * @param p2 Additional data for the command (see #CommandProc)
- * @param cmd The command to execute (a CMD_* value)
- * @param callback A callback function to call after the command is finished
- * @param text The text to pass
- * @param my_cmd indicator if the command is from a company or server (to display error messages for a user)
- * @param estimate_only whether to give only the estimate or also execute the command
- * @return the command cost of this function.
- */
-CommandCost DoCommandPInternal(TileIndex tile, uint32 p1, uint32 p2, uint32 cmd, CommandCallback *callback, const std::string &text, bool my_cmd, bool estimate_only)
+bool DoCommandHelperBase::InternalExecutePrepTest(CommandFlags cmd_flags, TileIndex tile, Backup<CompanyID> &cur_company)
 {
-	/* Prevent recursion; it gives a mess over the network */
-	assert(_docommand_recursive == 0);
-	_docommand_recursive = 1;
-
-	/* Reset the state. */
-	_additional_cash_required = 0;
-
-	/* Get pointer to command handler */
-	byte cmd_id = cmd & CMD_ID_MASK;
-	assert(cmd_id < lengthof(_command_proc_table));
-
-	CommandProc *proc = _command_proc_table[cmd_id].proc;
-	/* Shouldn't happen, but you never know when someone adds
-	 * NULLs to the _command_proc_table. */
-	assert(proc != nullptr);
-
-	/* Command flags are used internally */
-	CommandFlags cmd_flags = GetCommandFlags(cmd);
-	/* Flags get send to the DoCommand */
-	DoCommandFlag flags = CommandFlagsToDCFlags(cmd_flags);
-
-	/* Make sure p2 is properly set to a ClientID. */
-	assert(!(cmd_flags & CMD_CLIENT_ID) || p2 != 0);
-
 	/* Do not even think about executing out-of-bounds tile-commands */
-	if (tile != 0 && (tile >= MapSize() || (!IsValidTile(tile) && (cmd_flags & CMD_ALL_TILES) == 0))) return_dcpi(CMD_ERROR);
+	if (tile != 0 && (tile >= MapSize() || (!IsValidTile(tile) && (cmd_flags & CMD_ALL_TILES) == 0))) return false;
```

Comment:
Doesn't `DoCommandHelper::Do` already perform this check?

Reply:
Yes, but this is for the `DoCommandHelper::Post/Execute` code path and is not called for `Do`.